lms/djangoapps/completion/models.py

```diff
@@ -0,0 +1,72 @@
+"""
+Completion tracking and aggregation models.
+"""
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from django.contrib.auth.models import User
+from django.core.exceptions import ValidationError
+from django.db import models
+from django.utils.translation import ugettext as _
+from model_utils.models import TimeStampedModel
+
+from openedx.core.djangoapps.xmodule_django.models import CourseKeyField, UsageKeyField
+
+
+def validate_ratio(value):
+    """
+    Verify that the passed value is between 0.0 and 1.0.
+    """
+    if not 0.0 <= value <= 1.0:
+        raise ValidationError(_('{value} must be between 0.0 and 1.0').format(value=value))
+
+
+class BlockCompletionManager(models.Manager):
+    """
+    Custom manager for BlockCompletion model.
+
+    Adds update_or_create method.
+    """
+
+    def update_or_create(self, user, course_key, block_key, completed):
+        """
+        Update the completed value for the specified record.
+        """
+
+        obj, isnew = self.get_or_create(
+            user=user,
+            course_key=course_key,
+            block_type=block_key.block_type,
+            block_key=block_key,
+            defaults={'completed': completed},
+        )
+        if not isnew and obj.completed != completed:
+            obj.completed = completed
+            obj.full_clean()
+            obj.save()
+        return obj
+
+
+class BlockCompletion(TimeStampedModel, models.Model):
+    """
+    Track completion of completable blocks.
+    """
+    user = models.ForeignKey(User)
+    course_key = CourseKeyField(max_length=255)
+    block_key = UsageKeyField(max_length=255)
+    block_type = models.CharField(max_length=64)
+    completed = models.FloatField(validators=[validate_ratio])
+
+    objects = BlockCompletionManager()
+
+    class Meta(object):
```

Comment:
Should we also include a composite index (`index_together`) on `user, course_key`? I imagine that retrieving all the data for a particular user in a particular course will be a common use case. But we can also wait to optimize/index later based on actual usage.

Reply:
No need.  MySQL can use the prefix of an index for indexing.  One of my favorite resources on how indexes work is [this five question quiz](http://use-the-index-luke.com/3-minute-test/mysql).  After you finish the quiz, it gives very clear explanations for why the answers are right or wrong, which makes it a great learning resource.

Reply:
Also, I think we will need to get the indexes right before this goes to production, because this table is going to get very large, and once it does, unless I'm mistaken about how MySQL does it, adding an index will be a prohibitively slow operation.